### ISSUE TITLE
LoRA Training: travail entamé, reste à finaliser

### ISSUE THREAD

**Issue by VincentGourbin** [VincentGourbin/ltx-video-swift-mlx#1]
## Contexte

L'infrastructure de base pour l'entraînement LoRA sur Apple Silicon via MLX a été implémentée (commit 5ac66be+). Voici l'état des lieux et le travail restant.

### Nouveaux fichiers créés (Phase 1-4)

| Fichier | Rôle |
|---------|------|
| `Sources/LTXVideo/Training/LoRALinear.swift` | Couche LoRA entraînable (A/B matrices) wrappant un `Linear` gelé |
| `Sources/LTXVideo/Training/LoRAInjector.swift` | Injection LoRA dans le transformer (pattern identique à `quantize()` de MLX) |
| `Sources/LTXVideo/Training/LoRATrainingConfig.swift` | Config avec presets mémoire (compact/balanced/quality/max) |
| `Sources/LTXVideo/Training/FlowMatchingLoss.swift` | Loss velocity-prediction MSE (vidéo + audio dual) |
| `Sources/LTXVideo/Training/VideoDataset.swift` | Chargement vidéos + captions + audio via AVFoundation |
| `Sources/LTXVideo/Training/LatentCache.swift` | Cache disque latents VAE + embeddings texte (safetensors) |
| `Sources/LTXVideo/Training/TrainingLoop.swift` | Boucle d'entraînement complète avec `valueAndGrad` + `AdamW` |
| `Sources/LTXVideo/Training/LoRASaver.swift` | Sauvegarde format ComfyUI/Diffusers compatible |
| `Sources/LTXVideoCLI/TrainCommand.swift` | Sous-commande `ltx-video train` |

### Fichiers modifiés

| Fichier | Modification |
|---------|-------------|
| `Package.swift` | Ajout `MLXOptimizers` |
| `Sources/LTXVideo/LoRA/LoRALoader.swift` | Ajout `LoRAKeyMapper.modelKeyToLoraKey()` (mapping inverse) |
| `Sources/LTXVideo/Pipeline/LTXPipeline.swift` | APIs training: `getTransformerForTraining()`, `encodeVideoLatents()`, `encodeAudioLatents()`, `encodeAudioText()`, `TransformerRef` |
| `Sources/LTXVideoCLI/LTXVideoCLI.swift` | Ajout `Train.self` subcommand |

### Compile et CLI fonctionne

```bash
swift build  # ✅ 0 errors, 0 warnings
ltx-video train --help  # ✅ affiche toutes les options
swift test  # ✅ tests passent
```

### Phase 2 : Validation données

- [ ] Test end-to-end dataset loading (vidéo → frames → MLXArray)
- [ ] Test round-trip latent cache (encode → save → reload → identique)
- [ ] AudioVAE encoder (actuellement placeholder zeros — `AudioVAE` n'a que `decode()`)

### Phase 3 : Validation round-trip LoRA

- [ ] Train 50 steps → save → reload avec `LoRALoader` existant → inférence OK
- [ ] Valider que le format safetensors est compatible ComfyUI/Python
- [ ] Test avec `pipeline.fuseLoRA(from: saved_path)` → vidéo générée sans crash

### Phase 4 : Tests réels

- [ ] Overfit sur 1-3 vidéos pour vérifier que loss → 0
- [ ] Vérifier que `scaledDotProductAttention` supporte les gradients en MLX
  - Risque : certains ops MLX n'ont pas de backward. Fallback manual attention si nécessaire.
- [ ] Mesurer mémoire peak sur différentes configs (256x256x9 int4 vs 512x512x25 bf16)

### Phase 5 : Optimisations

- [ ] QLoRA (base int4/qint8 + LoRA float32) — vérifier que `valueAndGrad` fonctionne avec `QuantizedLinear` base
- [ ] Gradient accumulation multi-step
- [ ] LR scheduling cosine avec warmup
- [ ] Validation loop (génère vidéo sample tous les N steps)
- [ ] EMA (Exponential Moving Average)

## ⚠️ Mis en pause

Ce travail est **mis en pause** en faveur de la compatibilité LTX-2.3 (voir issue #2). Le code de training devra être adapté aux changements architecturaux de LTX-2.3 une fois la migration effectuée.

## CLI Usage (prévu)

```bash
ltx-video train ./my-dataset \
    --output ./my-lora --model dev --rank 16 --alpha 16 \
    --lr 1e-4 --steps 2000 --save-every 250 \
    -w 256 -h 256 -f 9 \
    --transformer-quant int4 --audio --audio-loss-weight 0.5
```

**Comment by VincentGourbin** [comment]
## État des lieux — juillet 2026 (branche `feature/lora-training-validation`)

Beaucoup des cases de 2026-03 sont faites depuis longtemps, mise à jour :

### Fait depuis l'ouverture ✅
- Checkpoints + resume + pause/stop (`TrainingController`, `TrainingState`)
- Gradient accumulation (`--grad-accum`), clipping (`--max-grad-norm`), warmup
- Preview vidéo aux checkpoints (`--preview-prompt`, subprocess distilled)
- Encodeur AudioVAE réel (`includeEncoder: true`) — plus de placeholder zeros
- Format safetensors ComfyUI/PEFT validé byte-for-byte (`LoRADeltaParityTests`, mai 2026)
- `scaledDotProductAttention` backward : OK (le training tourne)

### Nouveau sur cette branche (bumps mlx-swift 0.31.x) 🆕
- **Schedulers officiels** (`MLXOptimizers` 0.31.5) : warmup linéaire → **cosine decay** vers 10% du pic (`--lr-schedule cosine|constant`, défaut cosine)
- **AdamW bias correction** (0.31.4) activée — aligne sur torch.optim.AdamW
- **Instrumentation mémoire** : `TrainingProgress.peakMemoryBytes` + peak GPU loggé par step et en fin de run
- Note : le gradient checkpointing (remat) n'est **pas** exposé par mlx-swift — le levier mémoire est le QLoRA (base quantifiée gelée), en cours de validation

### Validation en cours sur cette branche 🔄
- Dataset : `Wild-Heart/Disney-VideoGeneration-Dataset` (Apache-2.0, 69 clips Steamboat Willie) converti au format mp4+txt
- [ ] Overfit 3 clips × 200 steps (loss → ~0) — **en cours**
- [ ] QLoRA : runs courts qint8 / int4, mémoire peak comparée à bf16
- [ ] Round-trip : train → save → `fuseLoRA` → génération
- [ ] Run de style complet sur les 69 clips (nécessite le modèle `dev`, ~46 GB à télécharger — à décider)

### Restera après ⏭
- EMA (optionnel)
- Training audio (`--audio` toujours gated off)

**Comment by VincentGourbin** [comment]
## Résultats de la campagne de validation (PR #38)

- ✅ **Overfit 3 clips × 200 steps** : loss lissée en décroissance monotone 0,686 → 0,564 (best 0,407) — la machinerie apprend.
- ✅ **QLoRA validé** : qint8 = parité de loss avec bf16 à 4 décimales (même seed/step), **peak 43,6 GB vs 84,3 GB** ; int4 = 37,5 GB, plus bruité mais apprend. Le bf16 sur 96 GB **swappe** (~110 s/step) → `--transformer-quant qint8` est la config recommandée.
- ✅ **Round-trip complet** : train → save → `fuseLoRA` (384 layer-pairs) → génération sans erreur.
- ✅ Schedulers officiels (cosine+warmup), AdamW bias-correction, instrumentation mémoire peak.
- 📚 Documenté dans le bundle OKF (décision QLoRA + baselines).

Reste sur cette issue : le run de style complet 69 clips (idéalement sur `dev`, ~46 GB à télécharger), EMA optionnel, training audio.

**Comment by VincentGourbin** [comment]
## Run de style complet : terminé ✅

69 clips Disney (Steamboat Willie), base **dev + qint8**, rank 32, 49 frames, trigger `STMBTW` :

- **320×192×49 : 1500 steps en 4 h 45, peak 61,2 GB constant du step 1 au step 1500**, best loss 0,187, 6 checkpoints + LoRA final sauvés.
- Leçon mémoire chèrement acquise : 448×256×49 (784 tokens) meurt par jetsam à 95,4 GB — la règle de dimensionnement (~60 MB/token d'entraînement au-dessus de la base qint8, pas de remat dans mlx-swift) est documentée dans `docs/knowledge/benchmarks/lora-training-baselines-m3max.md`.
- Génération A/B avec trigger livrée pour jugement visuel.

Avec la PR #38 (schedulers, bias-correction, instrumentation, QLoRA validé), cette issue peut probablement se fermer — resteraient EMA (optionnel) et le training audio, à re-ticketer séparément si besoin.

**Comment by VincentGourbin** [comment]
Fermée : le training LoRA est validé de bout en bout et les modernisations sont mergées (PR #38 — schedulers cosine officiels, AdamW bias-correction, instrumentation mémoire, défaut QLoRA qint8, garde-fous resume/warmup/quants). Preuves et baselines : docs/knowledge (décision QLoRA + baselines M3 Max). LoRA de style Disney produit et vérifié en génération. Restent hors périmètre de cette issue : EMA (optionnel) et le training audio — à ré-ouvrir en issues dédiées si le besoin se présente.
